core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+            'class_to_answer_group_mapping': self.class_to_answer_group_mapping,
+            'cv_vocabulary': self.count_vector.__dict__['vocabulary_']
+        }
+        return classifier_data
+
+    # pylint: disable=too-many-locals
+    def train(self, training_data):
+        """Trains classifier using given training_data.
+
+        Args:
+            training_data: list(dict). The training data that is used for
+                training the classifier. The list contains dicts where each dict
+                represents a single training data group, for example:
+                training_data = [
+                    {
+                        'answer_group_index': 1,
+                        'answers': ['a1', 'a2']
+                    },
+                    {
+                        'answer_group_index': 2,
+                        'answers': ['a2', 'a3']
+                    }
+                ]
+        """
+        # Answer group index to classifier class mapping.
+        answer_group_to_class_mapping = {
+            training_data[i]['answer_group_index']: i
+            for i in range(len(training_data))
+        }
+
+        class_to_answer_group_mapping = dict(zip(
+            answer_group_to_class_mapping.values(),
+            answer_group_to_class_mapping.keys()))
+
+        data = {}
+        count = 0
+        for answer_group in training_data:
+            answer_group_index = answer_group['answer_group_index']
+            for answer in answer_group['answers']:
+                data[count] = {
+                    'source': answer,
+                    'class': answer_group_to_class_mapping[answer_group_index]
+                }
+                count += 1
+
+        data, token_to_id = tokenize_data(data)
+
+        # No. of nearest neighbours to consider for classification using KNN.
+        top = len(answer_group_to_class_mapping) / 2
+
+        # GRID search for best value of T and K. T and K are model parameters
+        # of winnowing algorithm.
+        results = []
+        for T in range(T_MIN, T_MAX):
+            for K in range(K_MIN, T + 1):
+                data = generate_k_gram_hashes(data, token_to_id, K)
+                data = generate_program_fingerprints(data, T, K)
+                occurrence, missclassified_points = run_knn(data, top)
+                accuracy = (
+                    len(data) - len(missclassified_points)) / float(len(data))
+                results.append((accuracy, T, K))
+
+        best_score = sorted(results, key=lambda e: e[0], reverse=True)[0]
+        T = best_score[1]
+        K = best_score[2]
+
+        # Run KNN for best value of T and K.
+        data = generate_k_gram_hashes(data, token_to_id, K)
+        data = generate_program_fingerprints(data, T, K)
+        occurrence, missclassified_points = run_knn(data, top)
+
+        # Create a new dataset for missclassified points.
+        programs = [data[pid]['source'] for pid in sorted(data.keys())]
```

Comment:
Why would keys be already sorted? Just because keys are added in ascending order does not mean that dictionary will store them in sorted order. Dict stores them using a hash function. So it is not necessary for internal structure of dict to have same order as the order in which keys were added.